File: apps/frontend/src/adapters/shared/option-strategy.ts
import type {
  StrategyOverride,
  NewStrategyOverride,
  UpdateStrategyOverride,
} from "@/lib/types";

import { invoke, logger } from "./platform";

export const getOptionStrategyOverrides = async (
  accountIds: string[],
): Promise<StrategyOverride[]> => {
  try {
    return await invoke<StrategyOverride[]>("get_option_strategy_overrides", { accountIds });
  } catch (error) {
    logger.error("Error fetching option strategy overrides.");
    throw error;
  }
};

export const createOptionStrategyOverride = async (
  payload: NewStrategyOverride,
): Promise<StrategyOverride> => {
  try {
    return await invoke<StrategyOverride>("create_option_strategy_override", { payload });
  } catch (error) {
    logger.error("Error creating option strategy override.");
    throw error;
  }
};

export const updateOptionStrategyOverride = async (
  id: string,
  payload: UpdateStrategyOverride,
): Promise<StrategyOverride> => {
  try {
    return await invoke<StrategyOverride>("update_option_strategy_override", { id, payload });
  } catch (error) {
    logger.error("Error updating option strategy override.");
    throw error;
  }
};

export const deleteOptionStrategyOverride = async (id: string): Promise<void> => {
  try {
    await invoke<void>("delete_option_strategy_override", { id });
  } catch (error) {
    logger.error("Error deleting option strategy override.");
    throw error;
  }
};

File: apps/server/src/api/option_strategy.rs
use axum::{
    extract::{Path, RawQuery, State},
    http::StatusCode,
    routing::{delete, get, post, put},
    Json, Router,
};
use std::sync::Arc;
use wealthfolio_core::option_strategy::{
    NewStrategyOverride, StrategyOverride, UpdateStrategyOverride,
};

use crate::error::ApiResult;
use crate::main_lib::AppState;

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/option-strategy-overrides", get(get_option_strategy_overrides))
        .route("/option-strategy-overrides", post(create_option_strategy_override))
        .route(
            "/option-strategy-overrides/{id}",
            put(update_option_strategy_override),
        )
        .route(
            "/option-strategy-overrides/{id}",
            delete(delete_option_strategy_override),
        )
}

async fn get_option_strategy_overrides(
    State(state): State<Arc<AppState>>,
    RawQuery(raw): RawQuery,
) -> ApiResult<Json<Vec<StrategyOverride>>> {
    let account_ids: Vec<String> = raw
        .as_deref()
        .map(|q| {
            q.split('&')
                .filter_map(|p| p.strip_prefix("accountIds="))
                .map(|v| v.to_string())
                .collect()
        })
        .unwrap_or_default();
    let overrides = state
        .option_strategy_service
        .list_for_accounts(&account_ids)?;
    Ok(Json(overrides))
}

async fn create_option_strategy_override(
    State(state): State<Arc<AppState>>,
    Json(payload): Json<NewStrategyOverride>,
) -> ApiResult<Json<StrategyOverride>> {
    let created = state.option_strategy_service.create(payload).await?;
    Ok(Json(created))
}

async fn update_option_strategy_override(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Json(payload): Json<UpdateStrategyOverride>,
) -> ApiResult<Json<StrategyOverride>> {
    let updated = state.option_strategy_service.update(&id, payload).await?;
    Ok(Json(updated))
}

async fn delete_option_strategy_override(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> ApiResult<StatusCode> {
    state.option_strategy_service.delete(&id).await?;
    Ok(StatusCode::NO_CONTENT)
}
